pkg/maps/metricsmap/metricsmap.go

```diff
@@ -137,47 +137,134 @@ func (vs Values) Bytes() uint64 {
 	return b
 }
 
-func updateMetric(getCounter func() (prometheus.Counter, error), newValue float64) {
-	counter, err := getCounter()
+// metricsMapCollector implements Prometheus Collector interface
+type metricsmapCollector struct {
+	droppedCountDesc *prometheus.Desc
+	droppedByteDesc  *prometheus.Desc
+	forwardCountDesc *prometheus.Desc
+	forwardByteDesc  *prometheus.Desc
+
+	// eBPF code seems to expose multiple reasons for forwarded metrics
+	// as opposed to what is stated in bpf/lib/metrics.h comments.
+	// IterateWithCallback iterates through BPF map for each reason and direction.
+	// Since we do not have "reason" label on forwarded metrics, we would end up collecting
+	// same forwarded metric multiple times which is not allowed by prometheus client.
+	// See https://github.com/prometheus/client_golang/issues/242
+	//
+	// forwarded and dropped metrics maps are used to sum all values by desired set of labels
+	forwardedMetricsMap map[forwardLabels]metricValues
+	droppedMetricsMap   map[dropLabels]metricValues
+}
+
+type forwardLabels struct {
+	direction string
+}
+
+type dropLabels struct {
+	direction string
+	reason    string
+}
+
+type metricValues struct {
+	bytes float64
+	count float64
+}
+
+func newMetricsMapCollector() prometheus.Collector {
+	return &metricsmapCollector{
+		droppedMetricsMap:   make(map[dropLabels]metricValues),
+		forwardedMetricsMap: make(map[forwardLabels]metricValues),
+		droppedByteDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "drop_bytes_total"),
+			"Total dropped bytes, tagged by drop reason and ingress/egress direction",
+			[]string{metrics.LabelDropReason, metrics.LabelDirection}, nil,
+		),
+		droppedCountDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "drop_count_total"),
+			"Total dropped packets, tagged by drop reason and ingress/egress direction",
+			[]string{metrics.LabelDropReason, metrics.LabelDirection}, nil,
+		),
+		forwardCountDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "forward_count_total"),
+			"Total forwarded packets, tagged by ingress/egress direction",
+			[]string{metrics.LabelDirection}, nil,
+		),
+		forwardByteDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, metrics.SubsystemDatapath, "forward_bytes_total"),
+			"Total forwarded bytes, tagged by ingress/egress direction",
+			[]string{metrics.LabelDirection}, nil,
+		),
+	}
+}
+
+func (mc *metricsmapCollector) Collect(ch chan<- prometheus.Metric) {
+
```

Comment:
Nit: unnecessary newline

Reply:
@christarazi Thanks for the review. I didn't know that maps have a BatchLookup method; I'll keep this in mind in the future. As for the newline, I was relying on gomt to do the formatting, but I completely agree that it would be better without it here.